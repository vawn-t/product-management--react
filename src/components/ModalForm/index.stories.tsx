// Libraries
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import ModalForm from './index';

// Styles
import '@assets/styles/App.css';
import '@assets/styles/variables.css';
import { ProductTypes } from '@constants/types';

export default {
  title: 'Components/ModalForm',
  component: ModalForm
} as ComponentMeta<typeof ModalForm>;

const Template: ComponentStory<typeof ModalForm> = (args) => (
  <ModalForm {...args} />
);

export const Default = Template.bind({});
Default.args = {
  product: {
    id: '',
    imageUrl: '',
    name: 'Product',
    price: 10000,
    type: ProductTypes.Laptop
  },
  isModalShow: true,
  onCloseModal: () => undefined,
  onSubmitForm: () => undefined
};
