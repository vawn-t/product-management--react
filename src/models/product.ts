export interface Product {
  readonly id: string;
  name: string;
  type: string;
  price: number;
  imageUrl: string;
}
