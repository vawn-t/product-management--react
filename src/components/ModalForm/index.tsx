// Libraries
import React, { memo } from 'react';

// Components
import { Form, Modal } from '@components/index';

// Constants
import { PRODUCT_TYPE_LIST } from '@constants/index';

// Models
import { Product } from '@models/product';

// Types
import { FormVariants } from '@common-types/index';

export interface ModalFormProps {
  product: Product;
  isModalShow: boolean;
  onSubmitForm: (product: Product) => void;
  onCloseModal: () => void;
}

const ModalForm: React.FC<ModalFormProps> = ({
  product,
  isModalShow,
  onSubmitForm,
  onCloseModal
}) => (
  <Modal isVisible={isModalShow} onClose={onCloseModal}>
    <Form
      variants={!product.id ? FormVariants.Create : FormVariants.Edit}
      options={PRODUCT_TYPE_LIST}
      productItem={product}
      onSubmit={onSubmitForm}
    />
  </Modal>
);

export default memo(ModalForm);
