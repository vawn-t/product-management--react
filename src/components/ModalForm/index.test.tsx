// Libraries
import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { create } from 'react-test-renderer';
import '@testing-library/jest-dom';

// Components
import ModalForm, { ModalFormProps } from './index';

// Constants
import { ProductTypes } from '@constants/index';

// Model
import { Product } from '@models/product';

describe('ModalForm component', () => {
  const defaultProps: ModalFormProps = {
    product: {
      id: '1',
      imageUrl: 'url',
      name: 'Product test',
      price: 1000000,
      type: ProductTypes.Phone
    },
    isModalShow: true,
    onCloseModal: jest.fn(),
    onSubmitForm: jest.fn()
  };

  let container: HTMLElement;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultTree = create(<ModalForm {...defaultProps} />).toJSON();
    expect(defaultTree).toMatchSnapshot();
  });

  test('should submitMock be called', () => {
    const submitMock = jest.fn();

    render(<ModalForm {...defaultProps} onSubmitForm={submitMock} />, {
      container
    });

    const button: HTMLButtonElement = screen.getByRole('button');

    fireEvent.click(button);

    expect(submitMock).toHaveBeenCalled();
  });

  test('should handleClose be called', () => {
    const closeMock = jest.fn();

    render(<ModalForm {...defaultProps} onCloseModal={closeMock} />),
      { container };

    const img: HTMLImageElement = screen.getByRole('img');

    fireEvent.click(img);

    expect(closeMock).toHaveBeenCalled();
  });
});
