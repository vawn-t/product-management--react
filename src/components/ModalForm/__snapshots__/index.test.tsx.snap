// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalForm component should render correctly 1`] = `
<div
  className="modal"
>
  <div
    className="modal-dialog"
  >
    <div
      className="close-icon"
    >
      <img
        alt="close icon"
        className="image"
        onClick={[MockFunction]}
      />
    </div>
    <div
      className="modal-body"
    >
      <div
        className="form-wrapper"
      >
        <div
          className="title"
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
              "fontStyle": "inherit",
              "fontWeight": undefined,
              "lineHeight": undefined,
            }
          }
        >
          Edit
           Product
        </div>
        <form
          className="form"
          name="form"
          onSubmit={[Function]}
        >
          <fieldset
            className="field-group"
          >
            <div
              className="field-wrapper"
            >
              <label
                htmlFor="name"
              >
                Product name
                : 
              </label>
              <div
                className="input-wrapper"
              >
                <input
                  className="field field-standard"
                  disabled={false}
                  id="name"
                  name="name"
                  onChange={[Function]}
                  placeholder="Enter productItem name..."
                  readOnly={false}
                  required={false}
                  style={
                    Object {
                      "height": undefined,
                      "width": undefined,
                    }
                  }
                  type="text"
                  value="Product test"
                />
              </div>
            </div>
            <div
              className="validate-message"
            >
              <p
                className="text-highlight"
                style={
                  Object {
                    "color": "var(--danger)",
                    "fontSize": "12px",
                    "fontStyle": "inherit",
                    "fontWeight": 400,
                  }
                }
              />
            </div>
            <div
              className="select-wrapper"
            >
              <label
                htmlFor="type"
              >
                Product type
                :
              </label>
              <select
                className="select"
                data-testid="select-option"
                disabled={false}
                id="type"
                name="type"
                onChange={[Function]}
                value="Phone"
              >
                <option
                  value=""
                >
                  Select
                </option>
                <option
                  value="Phone"
                >
                  Phone
                </option>
                <option
                  value="Tablet"
                >
                  Tablet
                </option>
                <option
                  value="Laptop"
                >
                  Laptop
                </option>
              </select>
            </div>
            <div
              className="validate-message"
            >
              <p
                className="text-highlight"
                style={
                  Object {
                    "color": "var(--danger)",
                    "fontSize": "12px",
                    "fontStyle": "inherit",
                    "fontWeight": 400,
                  }
                }
              />
            </div>
            <div
              className="field-wrapper"
            >
              <label
                htmlFor="price"
              >
                Price
                : 
              </label>
              <div
                className="input-wrapper"
              >
                <input
                  className="field field-standard"
                  disabled={false}
                  id="price"
                  name="price"
                  onChange={[Function]}
                  placeholder="Enter price..."
                  readOnly={false}
                  required={false}
                  style={
                    Object {
                      "height": undefined,
                      "width": undefined,
                    }
                  }
                  type="number"
                  value={1000000}
                />
              </div>
            </div>
            <div
              className="validate-message"
            >
              <p
                className="text-highlight"
                style={
                  Object {
                    "color": "var(--danger)",
                    "fontSize": "12px",
                    "fontStyle": "inherit",
                    "fontWeight": 400,
                  }
                }
              />
            </div>
            <div
              className="field-wrapper"
            >
              <label
                htmlFor="imageUrl"
              >
                Image link
                : 
              </label>
              <div
                className="input-wrapper"
              >
                <input
                  className="field field-standard"
                  disabled={false}
                  id="imageUrl"
                  name="imageUrl"
                  onChange={[Function]}
                  placeholder="Enter image link..."
                  readOnly={false}
                  required={false}
                  style={
                    Object {
                      "height": undefined,
                      "width": undefined,
                    }
                  }
                  type="text"
                  value="url"
                />
              </div>
            </div>
            <div
              className="validate-message"
            >
              <p
                className="text-highlight"
                style={
                  Object {
                    "color": "var(--danger)",
                    "fontSize": "12px",
                    "fontStyle": "inherit",
                    "fontWeight": 400,
                  }
                }
              />
            </div>
          </fieldset>
          <button
            className="btn btn-primary"
            disabled={false}
            onClick={[Function]}
          >
            Edit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
